### PR TITLE
MINOR: update release script

### PR DESCRIPTION
**Pull request by mjsax**
*More detailed description of your change,
if necessary. The PR title and PR message become
the squashed commit message, so use a separate
comment to ping reviewers.*

*Summary of testing strategy (including rationale)
for the feature or bug fix. Unit and/or integration
tests are expected for any behaviour change and
system tests should be considered for larger changes.*

### Committer Checklist (excluded from commit message)
- [ ] Verify design and implementation 
- [ ] Verify test coverage and CI build status
- [ ] Verify documentation (including upgrade notes)

### CODE REVIEW

**Review by guozhangwang — COMMENTED** [review]
Assuming this is just for cherry-picking the changes from trunk to 0.11.0, lgtm.

**Comment by mjsax** [comment]
@guozhangwang Some is from `trunk`, some is additional edits...

**Comment by guozhangwang** [comment]
@mjsax could you leave comments on where the additional edits are?

File: release.py [hunk 1]
```diff
@@ -45,8 +45,8 @@
 SCRIPT_DIR = os.path.abspath(os.path.dirname(__file__))
 # Location of the local git repository
 REPO_HOME = os.environ.get("%s_HOME" % CAPITALIZED_PROJECT_NAME, SCRIPT_DIR)
-# Remote name which points to Apache git
-PUSH_REMOTE_NAME = os.environ.get("PUSH_REMOTE_NAME", "apache")
+# Remote name, which points to Github by default
+PUSH_REMOTE_NAME = os.environ.get("PUSH_REMOTE_NAME", "apache-github")
```
Comment (mjsax) [inline release.py:49]: renaming of variable as we use Github now

File: release.py [hunk 2]
```diff
@@ -77,6 +77,7 @@ def print_output(output):
 def cmd(action, cmd, *args, **kwargs):
     if isinstance(cmd, basestring) and not kwargs.get("shell", False):
         cmd = cmd.split()
+    allow_failure = kwargs.pop("allow_failure", False)
```
Comment (mjsax) [inline release.py:80]: new

File: release.py [hunk 3]
```diff
@@ -93,6 +94,9 @@ def cmd(action, cmd, *args, **kwargs):
     except subprocess.CalledProcessError as e:
         print_output(e.output)
 
+        if allow_failure:
+            return
+
```
Comment (mjsax) [inline release.py:99]: new

File: release.py [hunk 4]
```diff
@@ -126,9 +130,9 @@ def sftp_mkdir(dir):
     try:
        cmd_str  = """
 cd %s
-mkdir %s
+-mkdir %s
```
Comment (mjsax) [inline release.py:133]: new (adding dash allows command to fail)

File: release.py [hunk 5]
```diff
@@ -126,9 +130,9 @@ def sftp_mkdir(dir):
...
 """ % (basedir, dirname)
-       cmd("Creating '%s' in '%s' in your Apache home directory if it does not exist (errors are ok if the directory already exists)" % (dirname, basedir), "sftp -b - %s@home.apache.org" % apache_id, stdin=cmd_str)
+       cmd("Creating '%s' in '%s' in your Apache home directory if it does not exist (errors are ok if the directory already exists)" % (dirname, basedir), "sftp -b - %s@home.apache.org" % apache_id, stdin=cmd_str, allow_failure=True)
```
Comment (mjsax) [inline release.py:135]: new

File: release.py [hunk 6]
```diff
@@ -162,6 +166,32 @@ def get_pref(prefs, name, request_fn):
       signing.keyId=your-gpgkeyId
       signing.password=your-gpg-passphrase
       signing.secretKeyRingFile=/Users/your-id/.gnupg/secring.gpg (if you are using GPG 2.1 and beyond, then this file will no longer exist anymore, and you have to manually create it from the new private key directory with "gpg --export-secret-keys -o ~/.gnupg/secring.gpg")
+8. ~/.m2/settings.xml configured for pgp signing and uploading to apache release maven, i.e.,
```
Comment (mjsax) [inline release.py:169]: c&p from `trunk`

File: release.py [hunk 7]
```diff
@@ -262,6 +291,12 @@ def select_gpg_key():
 print("Updating version numbers")
 replace("gradle.properties", "version", "version=%s" % release_version)
 replace("tests/kafkatest/__init__.py", "__version__", "__version__ = '%s'" % release_version)
+cmd("update streams quickstart pom", ["sed", "-i", ".orig"," s/-SNAPSHOT//", "streams/quickstart/pom.xml"])
+cmd("update streams quickstart java pom", ["sed", "-i", ".orig", "s/-SNAPSHOT//", "streams/quickstart/java/pom.xml"])
+cmd("update streams quickstart java pom", ["sed", "-i", ".orig", "s/-SNAPSHOT//", "streams/quickstart/java/src/main/resources/archetype-resources/pom.xml"])
+cmd("remove backup pom.xml", "rm streams/quickstart/pom.xml.orig")
+cmd("remove backup java pom.xml", "rm streams/quickstart/java/pom.xml.orig")
+cmd("remove backup java pom.xml", "rm streams/quickstart/java/src/main/resources/archetype-resources/pom.xml.orig")
```
Comment (mjsax) [inline release.py:299]: c&p from `trunk`

File: release.py [hunk 8]
```diff
@@ -277,6 +312,8 @@ def select_gpg_key():
 os.makedirs(work_dir)
 print("Temporary build working director:", work_dir)
 kafka_dir = os.path.join(work_dir, 'kafka')
+streams_quickstart_dir = os.path.join(kafka_dir, 'streams/quickstart')
+print("Streams quickstart dir", streams_quickstart_dir)
```
Comment (mjsax) [inline release.py:316]: c&p from `trunk`

File: release.py [hunk 9]
```diff
@@ -306,7 +343,7 @@ def select_gpg_key():
 
 cmd("Building artifacts", "gradle", cwd=kafka_dir, env=jdk7_env)
 cmd("Building artifacts", "./gradlew clean releaseTarGzAll aggregatedJavadoc", cwd=kafka_dir, env=jdk7_env)
-# This should be removed with KAFKA-4421
+# This should be removed when Java7 is dropped (cf. KAFKA-4421)
```
Comment (mjsax) [inline release.py:346]: new

File: release.py [hunk 10]
```diff
@@ -324,7 +361,7 @@ def select_gpg_key():
     dir, fname = os.path.split(full_path)
     cmd("Generating MD5 for " + full_path, "gpg --print-md md5 %s > %s.md5" % (fname, fname), shell=True, cwd=dir)
     cmd("Generating SHA1 for " + full_path, "gpg --print-md sha1 %s > %s.sha1" % (fname, fname), shell=True, cwd=dir)
-    cmd("Generating SHA2 for " + full_path, "gpg --print-md sha512 %s > %s.sha2" % (fname, fname), shell=True, cwd=dir)
+    cmd("Generating SHA512 for " + full_path, "gpg --print-md sha512 %s > %s.sha512" % (fname, fname), shell=True, cwd=dir)
```
Comment (mjsax) [inline release.py:364]: from `trunk`

File: release.py [hunk 11]
```diff
@@ -338,23 +375,24 @@ def select_gpg_key():
...
+""" % (local_path, remote_path)
+        cmd("Uploading artifacts in %s to your Apache home directory" % root, "sftp -b - %s@home.apache.org" % apache_id, stdin=sftp_cmds)
 
 with open(os.path.expanduser("~/.gradle/gradle.properties")) as f:
     contents = f.read()
 if not user_ok("Going to build and upload mvn artifacts based on these settings:\n" + contents + '\nOK (y/n)?: '):
     fail("Retry again later")
 cmd("Building and uploading archives", "./gradlew uploadArchivesAll", cwd=kafka_dir, env=jdk7_env)
 cmd("Building and uploading archives", "./gradlew uploadCoreArchives_2_12 -PscalaVersion=2.12", cwd=kafka_dir, env=jdk8_env)
+cmd("Building and uploading archives", "mvn deploy -Pgpg-signing", cwd=streams_quickstart_dir, env=jdk7_env)
```
Comment (mjsax) [inline release.py:395]: from `trunk`

File: release.py [hunk 12]
```diff
@@ -338,23 +375,24 @@ def select_gpg_key():
 for root, dirs, files in os.walk(artifacts_dir):
     assert root.startswith(artifacts_dir)
 
-    for file in files:
-        local_path = os.path.join(root, file)
-        remote_path = os.path.join("public_html", kafka_output_dir, root[len(artifacts_dir)+1:], file)
-        sftp_cmds += "\nput %s %s" % (local_path, remote_path)
-
```
Comment (mjsax) [inline release.py:345]: new (compare next below)

File: release.py [hunk 13]
```diff
@@ -338,23 +375,24 @@ def select_gpg_key():
...
     for dir in dirs:
         sftp_mkdir(os.path.join("public_html", kafka_output_dir, root[len(artifacts_dir)+1:], dir))
 
-if sftp_cmds:
```
Comment (mjsax) [inline release.py:349]: new (compare above)

This make the script more interactive as it reports progress during upload

File: release.py [hunk 14]
```diff
@@ -383,11 +421,11 @@ def select_gpg_key():
       wget http://home.apache.org/~%(apache_id)s/kafka-%(rc_tag)s/kafka-%(release_version)s-src.tgz.asc &&
       wget http://home.apache.org/~%(apache_id)s/kafka-%(rc_tag)s/kafka-%(release_version)s-src.tgz.md5 &&
       wget http://home.apache.org/~%(apache_id)s/kafka-%(rc_tag)s/kafka-%(release_version)s-src.tgz.sha1 &&
-      wget http://home.apache.org/~%(apache_id)s/kafka-%(rc_tag)s/kafka-%(release_version)s-src.tgz.sha2 &&
+      wget http://home.apache.org/~%(apache_id)s/kafka-%(rc_tag)s/kafka-%(release_version)s-src.tgz.sha512 &&
```
Comment (mjsax) [inline release.py:424]: from `trunk`

File: release.py [hunk 15]
```diff
@@ -383,11 +421,11 @@ def select_gpg_key():
...
       gpg --verify kafka-%(release_version)s-src.tgz.asc kafka-%(release_version)s-src.tgz &&
       gpg --print-md md5 kafka-%(release_version)s-src.tgz | diff - kafka-%(release_version)s-src.tgz.md5 &&
       gpg --print-md sha1 kafka-%(release_version)s-src.tgz | diff - kafka-%(release_version)s-src.tgz.sha1 &&
-      gpg --print-md sha512 kafka-%(release_version)s-src.tgz | diff - kafka-%(release_version)s-src.tgz.sha2 &&
+      gpg --print-md sha512 kafka-%(release_version)s-src.tgz | diff - kafka-%(release_version)s-src.tgz.sha512 &&
```
Comment (mjsax) [inline release.py:428]: from `trunk`

File: release.py [hunk 16]
```diff
@@ -439,7 +477,7 @@ def select_gpg_key():
 http://home.apache.org/~%(apache_id)s/kafka-%(rc_tag)s/javadoc/
 
 * Tag to be voted upon (off %(dev_branch)s branch) is the %(release_version)s tag:
-https://git-wip-us.apache.org/repos/asf?p=kafka.git;a=tag;h=%(rc_githash)s
+https://github.com/apache/kafka/releases/tag/%(rc_tag)s
```
Comment (mjsax) [inline release.py:480]: new

File: release.py [hunk 17]
```diff
@@ -449,7 +487,7 @@ def select_gpg_key():
 
 * Successful Jenkins builds for the %(dev_branch)s branch:
 Unit/integration tests: https://builds.apache.org/job/kafka-%(dev_branch)s-jdk7/<BUILD NUMBER>/
-System tests: https://jenkins.confluent.io/job/system-test-kafka-%(dev_branch)s/<BUILD_NUMBER>/
+System tests: https://jenkins.confluent.io/job/system-test-kafka/job/%(dev_branch)s/<BUILD_NUMBER>/
```
Comment (mjsax) [inline release.py:490]: new

**Review by mjsax — COMMENTED** [review]
Done.

I think the "new fixes" should to forward ported to all to all other branches, too.

File: release.py (hunk 9 above)
Comment (guozhangwang) [reply release.py:346]: this is not a comment: https://issues.apache.org/jira/browse/KAFKA-4421 is already merged in trunk, so we should modify this place if wanting to cherry pick to newer branches.

File: release.py (hunk 17 above)
Comment (guozhangwang) [reply release.py:490]: ack, thanks!

**Review by guozhangwang — APPROVED** [review]

File: release.py (hunk 9 above)
Comment (ewencp) [reply release.py:346]: 👍  was going to make the exact same comment

**Review by ewencp — APPROVED** [review]

File: release.py (hunk 9 above)
Comment (mjsax) [reply release.py:346]: Yes. This PR is for 0.11.0 and should similar for 1.0 and 1.1 -- for 2.0 and `trunk` we need to update the script accordingly.

I'll merge this PR as well as #5347 and #5328 and move on to forward port this to other branches.

**Pull request merged**